### ISSUE TITLE
`reassociate` creates loop dependent variables and defeats `licm`

### ISSUE THREAD

**Issue by qcolombet** [llvm/llvm-project#61458]
I stumbled on a case where the reassociate pass is actively making the generated code worse.
In a nutshell, reassociate turns loop invariant computations into loop variant ones and licm is not able to hoist them back.

I haven't dug why, maybe reassociate is tweaked to form cse'able expressions / reduce the total number of instructions but the overall runtime is going to be worse in this case.

To reproduce:
```shell
opt -S -passes='reassociate' repro_reassoc_defeats_licm.ll -o -
```

Result:
Before reassociate we have:
- 2 loop invariant adds
- 1 loop variant shift
- 3 loop variant adds

After reassociate we have:
- 0 loop invariant adds
- 1 loop variant shift
- 5 loop variant adds (of which 1 is cse'able)

So accounting for `cse` (`-passes=reassociate,early-cse`), we would trade 2 loop invariant adds + 3 loop variant adds with 4 loop variant adds.

Before:
```ll
define void @reassoc_defeats_licm(i64 %inv1, i64 %inv2, i64 %inv3) {
bb:
  %inv4 = add nuw nsw i64 %inv1, %inv2
  %inv5 = add nuw nsw i64 %inv2, %inv3
  br label %bb214

bb214:                                            ; preds = %bb214, %bb
  %iv1 = phi i64 [ %iv2, %bb214 ], [ 0, %bb ]
  %iv2 = phi i64 [ %iv2_plus_1, %bb214 ], [ 1, %bb ]
  %loop_dependent = shl nuw nsw i64 %iv1, 13
  %loop_dependent2 = add nsw i64 %inv4, %loop_dependent
  call void @keep_alive(i64 %loop_dependent2)
  %loop_dependent3 = add i64 %inv5, %loop_dependent
  call void @keep_alive(i64 %loop_dependent3)
  %iv2_plus_1 = add i64 %iv2, 1
  br label %bb214
}

declare void @keep_alive(i64)
```

After (with cse):
```ll
define void @reassoc_defeats_licm(i64 %inv1, i64 %inv2, i64 %inv3) {
bb:
  br label %bb214

bb214:                                            ; preds = %bb214, %bb
  %iv1 = phi i64 [ %iv2, %bb214 ], [ 0, %bb ]
  %iv2 = phi i64 [ %iv2_plus_1, %bb214 ], [ 1, %bb ]
  %loop_dependent = shl nuw nsw i64 %iv1, 13
  %inv4 = add i64 %loop_dependent, %inv2
  %loop_dependent2 = add i64 %inv4, %inv1
  call void @keep_alive(i64 %loop_dependent2)
  %loop_dependent3 = add i64 %inv4, %inv3
  call void @keep_alive(i64 %loop_dependent3)
  %iv2_plus_1 = add i64 %iv2, 1
  br label %bb214
}

declare void @keep_alive(i64)
```

[repro_reassoc_defeats_licm.ll.txt](https://github.com/llvm/llvm-project/files/10994658/repro_reassoc_defeats_licm.ll.txt)

**Comment by qcolombet** [comment]
Okay I think I made some progress on this.
As far as I can tell this behavior has been introduced with b8a330c42ab43879119cd3a305756d28aefe9fe6. This commit purposely reorders the expressions to expose more CSE opportunities.

If I skip the related code (with a stupid `&& 0` on this line https://github.com/llvm/llvm-project/blob/478bd0735fc094d3af37e9791df5118a402ae7a7/llvm/lib/Transforms/Scalar/Reassociate.cpp#L2402), the problematic reassociation doesn't happen and regular expression reassociations happen as expected.
Put differently I think we need to rework the heuristic used here.

**Comment by qcolombet** [comment]
Tentative fix at https://reviews.llvm.org/D147457.

I limited the scope of reassociation for CSE for expression that are local to the same basic block.
This is not the greatest fix in the world, but at least it is predictable. A proper way to fix that would be to come up with an actual cost model, but that needs to be done globally (all expressions at a time) and that's probably overkill and not worth it.

**Comment by qcolombet** [comment]
Was fixed with a4e88cba1837
